scripts/eurostat/life_expectancy/preprocess.py

```diff
@@ -0,0 +1,96 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pandas as pd
+import re
+
+PATH = 'demo_r_mlifexp.tsv'
+
+def nuts_to_iso(data):
+    """Convert 2 letter nuts code of countries to ISO Alpha-3 code"""
+    ISO_2_TO_3_PATH = ('https://gist.githubusercontent.com/tadast/8827699/raw/'
+                       '3cd639fa34eec5067080a61c69e3ae25e3076abb/'
+                       'countries_codes_and_coordinates.csv')
+    codes = pd.read_csv(ISO_2_TO_3_PATH)
+    codes["Alpha-2 code"] = codes["Alpha-2 code"].str.extract(r'"([a-zA-Z]+)"')
+    codes["Alpha-3 code"] = codes["Alpha-3 code"].str.extract(r'"([a-zA-Z]+)"')
+    # NUTS code matches ISO Alpha-2 with two exceptions 
+    codes["NUTS"] = codes["Alpha-2 code"]
+    codes.loc[codes["NUTS"] == "GR", "NUTS"] = "EL" 
+    codes.loc[codes["NUTS"] == "GB", "NUTS"] = "UK"
+    code_dict = codes.set_index('NUTS').to_dict()['Alpha-3 code']
+    data.loc[data.index,'geo'] = data['geo'].map(code_dict)
+    assert (~data['geo'].isnull()).all()
+    return data
+
+def preprocess(filepath):
+    """Preprocess the tsv file for importing into DataCommons."""
+    data = pd.read_csv(filepath, sep='\t')
+
+    # Concatenate data of different years from multiple columns into one column
+    identifier = 'unit,sex,age,geo\\time'
+    assert identifier in data.columns
+    years = list(data.columns.values)
+    years.remove(identifier)
+    data = pd.melt(data, id_vars=identifier, value_vars=years, 
+                   var_name='year', value_name='life_expectancy')
+
+    # Format string into desired format
+    data['year'] = data['year'].astype(int) # remove spaces, e.g. "2018 "
+    num_null = data[data['life_expectancy'].isin([': ', ': e'])].shape[0]
+    data['life_expectancy'] = data['life_expectancy'].str.extract("(\d+\.\d+|\d+)")
+    assert num_null == data['life_expectancy'].isnull().sum()
```

Comment:
This is clean code.  Think about whether the following might be more robust, e.g., when another note letter is added to the data.

Write a function that takes a string containing the cell's value and have it discard any note.

def obtain_value(entry):
  entry = entry.split(':', maxsplit=-1)[0]  # Discard notes.
 if not entry:
    return NA (substitute however Pandas represents NA)
  return float(entry)

Then apply it to every value using apply: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.apply.html

Reply:
Revised as suggested.